### PR TITLE
Albrja/mic 5649/mypy fraemwork engine

### PR DESCRIPTION
**Pull request by albrja**
### Fix mypy errors in framework/engine.py
- *Category*: Type hinting
- *JIRA issue*: https://jira.ihme.washington.edu/browse/MIC-5649

### CODE REVIEW

File: src/vivarium/framework/engine.py [hunk 1]
```diff
@@ -262,29 +277,29 @@ def step(self) -> None:
...
         self,
         backup_path: Path | None = None,
-        backup_freq: int | float = None,
+        backup_freq: int | float | None = None,
     ) -> None:
-        if backup_freq:
+        if backup_freq and backup_path:
             time_to_save = time() + backup_freq
-            while self.current_time < self._clock.stop_time:
+            while self.current_time < self._clock.stop_time:  # type: ignore [operator]
```
Comment (albrja) [inline src/vivarium/framework/engine.py:290]: This an the next ignore are the same issues we have dealt with when comparing two ClockTime objects.

File: src/vivarium/framework/engine.py [hunk 2]
```diff
@@ -262,29 +273,29 @@ def step(self) -> None:
                 self._clock.event_time,
             )
             self._logger.debug(f"Updating: {len(pop_to_update)}")
-            self.time_step_emitters[event](pop_to_update)
+            self.time_step_emitters[event](pop_to_update, None)
```
Comment (rmudambi) [inline src/vivarium/framework/engine.py:276]: Same question here

File: src/vivarium/framework/engine.py [hunk 3]
```diff
@@ -343,13 +354,13 @@ def add_components(self, component_list: list[Component]) -> None:
         self._component_manager.add_components(component_list)
 
     def get_population(self, untracked: bool = True) -> pd.DataFrame:
-        return self._population.get_population(untracked)
+        return pd.DataFrame(self._population.get_population(untracked))
```
Comment (rmudambi) [inline src/vivarium/framework/engine.py:357]: Why is this necessary? `get_population()` is already type-hinted to return a DataFrame.

File: src/vivarium/framework/engine.py [hunk 4]
```diff
@@ -343,13 +354,13 @@ def add_components(self, component_list: list[Component]) -> None:
...
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return f"SimulationContext({self.name})"
 
     def get_number_of_steps_remaining(self) -> int:
-        return self._clock.time_steps_remaining
+        return int(self._clock.time_steps_remaining)
```
Comment (rmudambi) [inline src/vivarium/framework/engine.py:363]: Same question here.

File: src/vivarium/framework/engine.py [hunk 5]
```diff
@@ -365,64 +376,72 @@ class Builder:
...
 
-    def __init__(self, configuration: LayeredConfigTree, plugin_manager):
+    def __init__(
+        self, configuration: LayeredConfigTree, plugin_manager: PluginManager
+    ) -> None:
         self.configuration = configuration
         """Provides access to the :ref:`configuration<configuration_concept>`"""
 
-        self.logging: LoggingInterface = plugin_manager.get_plugin_interface("logging")
+        self.logging: LoggingInterface = plugin_manager.get_plugin_interface(LoggingInterface)
```
Comment (rmudambi) [inline src/vivarium/framework/engine.py:385]: The colons shouldn't be necessary anymore.

File: src/vivarium/framework/results/observer.py [hunk 6]
```diff
@@ -49,17 +51,17 @@ def get_configuration_name(self) -> str:
         return self.name.split("_observer")[0]
 
     @abstractmethod
-    def register_observations(self, builder: Builder) -> None:
+    def register_observations(self, builder: "Builder") -> None:
```
Comment (rmudambi) [inline src/vivarium/framework/results/observer.py:54]: You don't need quotes here.

**Review by patricktnast — APPROVED** [review]

File: src/vivarium/framework/engine.py [hunk 7]
```diff
@@ -241,7 +252,7 @@ def setup(self) -> None:
 
         post_setup = self._builder.event.get_emitter("post_setup")
         self._lifecycle.set_state("post_setup")
-        post_setup(None)
+        post_setup(pd.Index([]), None)
```
[the remark this answers is not in the document]
Comment (rmudambi) [reply src/vivarium/framework/engine.py:247]: For posterity, it's because this callable is `EventChannel.emit()` which takes two arguments - a pd.Index and an optional dictionary, with a default value of None.

This change is effectively changing the first (pd.Index) argument from None to an empty index, which is fine, because `post_setup` methods won't use the index anyway as the simulants haven't yet been initialized.

**Review by hussain-jafari — APPROVED** [review]
